src/diem/offchain/client.py

```diff
@@ -82,28 +82,36 @@ def send_request(
             raise CommandResponseError(cmd_resp)
         return cmd_resp
 
-    def process_inbound_request(self, request_sender_address: str, request_bytes: bytes) -> PaymentCommand:
+    def process_inbound_request(self, request_sender_address: str, request_bytes: bytes) -> CommandVariant:
         if not request_sender_address:
             raise protocol_error(ErrorCode.missing_http_header, f"missing {http_header.X_REQUEST_SENDER_ADDRESS}")
         try:
             _, public_key = self.get_base_url_and_compliance_key(request_sender_address)
         except ValueError as e:
             raise protocol_error(ErrorCode.invalid_x_request_sender_address, str(e)) from e
 
-        request = _deserialize_jws(request_bytes, CommandRequestObject, public_key, command_error)
-        if request.command_type == CommandType.PaymentCommand:
-            payment = typing.cast(PaymentCommandObject, request.command).payment
+        command_request_object = _deserialize_jws(request_bytes, CommandRequestObject, public_key, command_error)
+
+        if command_request_object.command_type == CommandType.PaymentCommand:
```

Comment:
isinstance(command_request_object, PaymentCommandObject), remove command_type abstract method.

it is better because the following typing.cast assumes it is PaymentCommandObject

Reply:
Actually I am not a fan of `typing.cast`. IMHO it is not so pythonic. Moreover, this is even the first time I see it used and I've seen my share of Python code hehe.

Reply:
if you remove the typing.cast, you get pyre checking error. because of code is expecting PaymentCommandObject, not other type.
I think the question is more of where we validate command_type == type of request.command.
When they are mismatched, what error will raise.

Reply:
I was going to add `isinstance` as you proposed and started with adding tests for the cases when the text in `CommandRequestObject.command_type` doesn't match the actual command object type. Was expecting the tests to fail but it looks like the code already generates descriptive exceptions even without using `isinstance`. The exceptions look like this:

```
        request = _deserialize_jws(request_bytes, CommandRequestObject, public_key, command_error)
        if request.command_type == CommandType.PaymentCommand:
            payment = typing.cast(PaymentCommandObject, request.command).payment
            self.validate_addresses(payment, request_sender_address)
            cmd = self.create_inbound_payment_command(request.cid, payment)
            if cmd.is_initial():
                self.validate_dual_attestation_limit(cmd.payment.action)
            elif cmd.is_rsend():
                self.validate_recipient_signature(cmd, public_key)
            return cmd
        elif request.command_type == CommandType.FundPullPreApprovalCommand:
            fund_pull_pre_approval = typing.cast(
>               FundPullPreApprovalCommandObject, request.command
            ).fund_pull_pre_approval
E           AttributeError: 'PaymentCommandObject' object has no attribute 'fund_pull_pre_approval'
```
Seems to me even nicer than adding `isinstance`. What do you think?